pkg/beacon/relay/entry/entry.go

```diff
@@ -33,26 +39,195 @@ func SignAndSubmit(
 	signer *dkg.ThresholdSigner,
 	startBlockHeight uint64,
 ) error {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	defer cancelCtx()
+
+	relayEntrySubmittedChannel := make(chan uint64)
+	subscription, err := relayChain.OnRelayEntrySubmitted(
+		func(event *event.EntrySubmitted) {
+			relayEntrySubmittedChannel <- event.BlockNumber
+		},
+	)
+	if err != nil {
+		return err
+	}
+	defer subscription.Unsubscribe()
+
+	chainConfig, err := relayChain.GetConfig()
+	if err != nil {
+		return err
+	}
+
+	relayEntryTimeoutChannel, err := blockCounter.BlockHeightWaiter(
+		startBlockHeight + chainConfig.RelayEntryTimeout,
+	)
+	if err != nil {
+		return err
+	}
+
 	previousEntry := new(bn256.G1)
-	_, err := previousEntry.Unmarshal(previousEntryBytes)
+	_, err = previousEntry.Unmarshal(previousEntryBytes)
 	if err != nil {
 		return err
 	}
 
-	initialState := &signatureShareState{
-		signingStateBase: signingStateBase{
-			channel:         channel,
-			relayChain:      relayChain,
-			blockCounter:    blockCounter,
-			signer:          signer,
-			previousEntry:   previousEntry,
-			honestThreshold: honestThreshold,
-		},
-		signingStartBlockHeight: startBlockHeight,
+	selfShare := signer.CalculateSignatureShare(previousEntry)
+
+	go broadcastShare(ctx, signer.MemberID(), selfShare, channel)
+
+	receiveChannel := make(chan net.Message, 64)
+	channel.Recv(ctx, func(netMessage net.Message) {
+		receiveChannel <- netMessage
+	})
+
+	receivedShares := map[group.MemberIndex]*bn256.G1{
+		signer.MemberID(): selfShare,
+	}
+
+	// Run the message loop until the number of received and valid signature
+	// shares is equal to the honest threshold. Message loop will be also
+	// terminated if an other member submits the result or the relay entry
+	// timeout block is reached.
+	for len(receivedShares) < honestThreshold {
+		select {
+		case netMessage := <-receiveChannel:
+			message, ok := netMessage.Payload().(*SignatureShareMessage)
+			if !ok || group.IsMessageFromSelf(signer.MemberID(), message) {
+				continue
+			}
+
+			share, err := extractShare(
```

Comment:
`extractAndValidateShare`? It took me a moment to find where the validation is 😆